client/src/client.rs

```diff
@@ -0,0 +1,254 @@
+///     .grpc_url(GRPC_URL)
+///     .plaintext_private_key("...")?
+///     .build()
+///     .await?;
+///
+/// let to_address = "celestia169s50psyj2f4la9a2235329xz7rk6c53zhw9mm".parse().unwrap();
+/// client.state().transfer(&to_address, 12345, TxConfig::default()).await?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// [`celestia-rpc`]: celestia_rpc
+/// [`celestia-grpc`]: celestia_grpc
+pub struct Client {
+    state: StateApi,
+    blob: BlobApi,
+    header: HeaderApi,
+    share: ShareApi,
+    fraud: FraudApi,
+    blobstream: BlobstreamApi,
+}
+
+/// A builder for [`Client`].
+#[derive(Debug, Default)]
+pub struct ClientBuilder {
+    rpc_url: Option<String>,
+    rpc_auth_token: Option<String>,
+    grpc_url: Option<String>,
+    pubkey: Option<VerifyingKey>,
+    signer: Option<DispatchedDocSigner>,
+}
+
+impl Context {
+    pub(crate) fn grpc(&self) -> Result<&TxClient<Transport, DispatchedDocSigner>> {
+        self.grpc.as_ref().ok_or(Error::ReadOnlyMode)
+    }
+
+    pub(crate) fn pubkey(&self) -> Result<&VerifyingKey> {
+        self.pubkey.as_ref().ok_or(Error::ReadOnlyMode)
+    }
+
+    pub(crate) async fn get_header_validated(&self, height: u64) -> Result<ExtendedHeader> {
+        let header = self.rpc.header_get_by_height(height).await?;
+        header.validate()?;
+        Ok(header)
+    }
+}
+
+impl Client {
+    /// Returns `ClientBuilder`.
+    pub fn builder() -> ClientBuilder {
+        ClientBuilder::new()
+    }
+
+    /// Returns state API accessor.
+    pub fn state(&self) -> &StateApi {
+        &self.state
+    }
+
+    /// Returns blob API accessor.
+    pub fn blob(&self) -> &BlobApi {
+        &self.blob
+    }
+
+    /// Returns blobstream API accessor.
+    pub fn blobstream(&self) -> &BlobstreamApi {
+        &self.blobstream
+    }
+
+    /// Returns header API accessor.
+    pub fn header(&self) -> &HeaderApi {
+        &self.header
+    }
+
+    /// Returns share API accessor.
+    pub fn share(&self) -> &ShareApi {
+        &self.share
+    }
+
+    /// Returns fraud API accessor.
+    pub fn fraud(&self) -> &FraudApi {
+        &self.fraud
+    }
+}
+
+impl ClientBuilder {
+    /// Returns a new builder.
+    pub fn new() -> ClientBuilder {
+        ClientBuilder::default()
+    }
+
+    /// Set signer and its public key.
+    pub fn singer<S>(mut self, pubkey: VerifyingKey, signer: S) -> ClientBuilder
+    where
+        S: DocSigner + Sync + Send + 'static,
+    {
+        self.pubkey = Some(pubkey);
+        self.signer = Some(DispatchedDocSigner::new(signer));
+        self
+    }
+
+    /// Set signer from a keypair.
+    pub fn keypair<S>(self, keypair: S) -> ClientBuilder
+    where
+        S: DocSigner + Keypair<VerifyingKey = VerifyingKey> + Sync + Send + 'static,
+    {
+        let pubkey = keypair.verifying_key();
+        self.singer(pubkey, keypair)
+    }
+
+    /// Set signer from a plaintext private key.
+    pub fn plaintext_private_key(self, s: &str) -> Result<ClientBuilder> {
```

Comment:
I'm thinking to rename this to `private_key_plaintext` and also add `private_key_bytes`. wdyt?

Reply:
Renamed to `private_key_plaintext` for hex string and added `private_key` for raw bytes.